File: lmlevelinfo/lm_tables.c
#include "lm_tables.h"

#include <stdio.h>
#include <string.h>

static void seterr(char *err, size_t cap, const char *msg) {
  if (!err || cap == 0) return;
  snprintf(err, cap, "%s", msg ? msg : "error");
}

static uint32_t snes(uint8_t bank, uint16_t addr) {
  return ((uint32_t)bank << 16) | (uint32_t)addr;
}

int lm_resolve_tables(const Rom *rom, LmTables *out, char *err, size_t errcap) {
  if (!rom || !rom->data || !out) {
    seterr(err, errcap, "lm_resolve_tables: invalid args");
    return 0;
  }
  memset(out, 0, sizeof(*out));

  // These pointer tables are literally never moved (per LevelTables.asm).
  out->layer1_ptr_table = snes(0x05, 0xE000);
  out->layer2_ptr_table = snes(0x05, 0xE600);
  out->sprite_ptr_table = snes(0x05, 0xEC00);

  // Secondary header base tables (vanilla fixed).
  out->sec_byte1 = snes(0x05, 0xF000);
  out->sec_byte2 = snes(0x05, 0xF200);
  out->sec_byte3 = snes(0x05, 0xF400);
  out->sec_byte4 = snes(0x05, 0xF600);

  // Optional sprite bank table: present if read1($0EF100) != $00
  // (LM adds a 512-byte table for the bank byte at $0EF100)
  uint8_t b = 0;
  if (rom_read8_snes(rom, snes(0x0E, 0xF100), &b) && b != 0x00) {
    out->sprite_bank_table = snes(0x0E, 0xF100);
    out->has_sprite_bank_table = 1;
  }

  // ---- Secondary header expansion (byte 5) ----
  // LevelTables.asm:
  // if read1($05D97D) == $22
  //   org $05<<16+read2(read3($05D97E)+5) ; $05DE00
  uint8_t op = 0;
  if (!rom_read8_snes(rom, snes(0x05, 0xD97D), &op)) {
    seterr(err, errcap, "Failed to read secondary header hijack probe");
    return 0;
  }
  if (op == 0x22) {
    uint32_t p = 0;
    uint16_t off = 0;
    if (rom_read24_snes(rom, snes(0x05, 0xD97E), &p) &&
        rom_read16_snes(rom, p + 5, &off)) {
      out->sec_byte5 = snes(0x05, (uint16_t)off);
      out->has_secondary_expansion = 1;
    }
  }

  // ---- Midway entrance hijack ----
  // LevelTables.asm:
  // if read1($05D9E3) == $22
  //   org read3(read3($05D9E4)+$0A) ; byte1
  //   org read3(read3($05D9E4)+$29) ; byte2
  //   org read3(read3($05D9E4)+$39) ; byte3
  if (rom_read8_snes(rom, snes(0x05, 0xD9E3), &op) && op == 0x22) {
    uint32_t p1 = 0;
    if (rom_read24_snes(rom, snes(0x05, 0xD9E4), &p1)) {
      uint32_t base = 0;
      if (rom_read24_snes(rom, p1 + 0x0A, &base)) out->midway_byte1 = base;
      if (rom_read24_snes(rom, p1 + 0x29, &base)) out->midway_byte2 = base;
      if (rom_read24_snes(rom, p1 + 0x39, &base)) out->midway_byte3 = base;
      if (out->midway_byte1 && out->midway_byte2 && out->midway_byte3) {
        out->has_midway_hijack = 1;
      }
    }
  }

  // ---- Secondary header extra bytes (LM 3.00+/3.40+) ----
  // From Level_Data_Format: additional per-level bytes at $06FA00, $06FC00, $06FE00.
  // These are fixed addresses when present. We treat them as present if mapping works and ROM is large enough.
  // If a hack is smaller than expected (rare), reads will fail and we leave them 0.
  uint8_t tmp = 0;
  if (rom_read8_snes(rom, snes(0x06, 0xFA00), &tmp)) out->sec_byte6 = snes(0x06, 0xFA00);
  if (rom_read8_snes(rom, snes(0x06, 0xFC00), &tmp)) out->sec_byte7 = snes(0x06, 0xFC00);
  if (rom_read8_snes(rom, snes(0x06, 0xFE00), &tmp)) out->sec_byte8 = snes(0x06, 0xFE00);

  return 1;
}

File: lib/jit-levels/levelinfo/lm-tables.js
/**
 * Lunar Magic table resolution — port of lmlevelinfo/lm_tables.c
 */

const { read8Snes, read16Snes, read24Snes, snesLoromToPc } = require('../smw-rom');
const { layer1BlobLooksValid } = require('./layer1-validate');

function snes(bank, addr) {
  return ((bank & 0xff) << 16) | (addr & 0xffff);
}

function pcToSnesLoromGuess(rom, pc) {
  if (pc >= rom.size) return null;
  let bank;
  let addr;
  const mapMode = rom.map_mode ?? rom.mapMode ?? 0x20;
  if (mapMode === 0x23) {
    if (rom.size > 0x400000 && pc >= 0x400000) {
      const pc2 = pc - 0x400000;
      bank = 0xc0 + Math.floor(pc2 / 0x10000);
      addr = pc2 % 0x10000;
    } else {
      bank = Math.floor(pc / 0x8000) & 0x3f;
      addr = 0x8000 + (pc % 0x8000);
    }
    return (bank << 16) | addr;
  }
  if (rom.size > 0x400000 && pc >= 0x400000) {
    const pc2 = pc - 0x400000;
    bank = 0x40 + Math.floor(pc2 / 0x8000);
    addr = 0x8000 + (pc2 % 0x8000);
  } else {
    bank = Math.floor(pc / 0x8000) & 0x7f;
    addr = 0x8000 + (pc % 0x8000);
  }
  if (bank > 0xff || addr > 0xffff) return null;
  return (bank << 16) | addr;
}

function pcToSnesHiromGuess(rom, pc) {
  if (pc >= rom.size) return null;
  const bank = 0xc0 + ((Math.floor(pc / 0x10000)) & 0x3f);
  const addr = pc % 0x10000;
  return (bank << 16) | addr;
}

function looksLikeLayer1BlobAtPtr(rom, ptrSnes24) {
  const pc = snesLoromToPc(rom, ptrSnes24);
  if (pc == null || pc + 6 >= rom.size) return false;
  return layer1BlobLooksValid(rom.data, pc, rom.size - pc);
}

function validateLayer1PtrTableBlock(rom, layer1TableSnes24) {
  let ok = 0;
  let total = 0;
  for (let i = 0; i < 0x200; i += 0x11) {
    const id = i & 0x1ff;
    const entry = layer1TableSnes24 + id * 3;
    const ptr = read24Snes(rom, entry);
    total++;
    if (ptr != null && looksLikeLayer1BlobAtPtr(rom, ptr)) ok++;
    if (total >= 32) break;
  }
  return ok >= 24;
}

function findLayer1PtrTableBlock(rom) {
  const step = 0x20;
  const maxPc = rom.size > 0x600 ? rom.size - 0x600 : 0;
  for (let pc = 0; pc <= maxPc; pc += step) {
    let snes24 = pcToSnesLoromGuess(rom, pc);
    if (snes24 != null && validateLayer1PtrTableBlock(rom, snes24)) return snes24;
    snes24 = pcToSnesHiromGuess(rom, pc);
    if (snes24 != null && validateLayer1PtrTableBlock(rom, snes24)) return snes24;
  }
  return null;
}

function lmResolveTables(rom) {
  if (!rom || !rom.data) {
    return { ok: false, error: 'lm_resolve_tables: invalid args' };
  }

  const tables = {
    layer1_ptr_table: snes(0x05, 0xe000),
    layer2_ptr_table: snes(0x05, 0xe600),
    sprite_ptr_table: snes(0x05, 0xec00),
    sprite_bank_table: 0,
    sec_byte1: snes(0x05, 0xf000),
    sec_byte2: snes(0x05, 0xf200),
    sec_byte3: snes(0x05, 0xf400),
    sec_byte4: snes(0x05, 0xf600),
    sec_byte5: 0,
    sec_byte6: 0,
    sec_byte7: 0,
    sec_byte8: 0,
    midway_byte1: 0,
    midway_byte2: 0,
    midway_byte3: 0,
    midway_byte4: 0,
    has_secondary_expansion: false,
    has_midway_hijack: false,
    has_midway_table4: false,
    has_sprite_bank_table: false,
  };

  if (!validateLayer1PtrTableBlock(rom, tables.layer1_ptr_table)) {
    const layer1Found = findLayer1PtrTableBlock(rom);
    if (layer1Found != null) {
      tables.layer1_ptr_table = layer1Found;
      tables.layer2_ptr_table = layer1Found + 0x600;
      tables.sprite_ptr_table = layer1Found + 0xc00;
      tables.sec_byte1 = layer1Found + 0x1000;
      tables.sec_byte2 = layer1Found + 0x1200;
      tables.sec_byte3 = layer1Found + 0x1400;
      tables.sec_byte4 = layer1Found + 0x1600;
    }
  }

  const spriteBankProbe = read8Snes(rom, snes(0x0e, 0xf100));
  if (spriteBankProbe != null && spriteBankProbe !== 0) {
    tables.sprite_bank_table = snes(0x0e, 0xf100);
    tables.has_sprite_bank_table = true;
  }

  const op = read8Snes(rom, snes(0x05, 0xd97d));
  if (op == null) {
    return { ok: false, error: 'Failed to read secondary header hijack probe' };
  }
  if (op === 0x22) {
    const p = read24Snes(rom, snes(0x05, 0xd97e));
    const off = p != null ? read16Snes(rom, p + 5) : null;
    if (p != null && off != null) {
      tables.sec_byte5 = snes(0x05, off);
      tables.has_secondary_expansion = true;
    }
  }

  const midwayOp = read8Snes(rom, snes(0x05, 0xd9e3));
  if (midwayOp === 0x22) {
    const p1 = read24Snes(rom, snes(0x05, 0xd9e4));
    if (p1 != null) {
      tables.midway_byte1 = read24Snes(rom, p1 + 0x0a) || 0;
      tables.midway_byte2 = read24Snes(rom, p1 + 0x29) || 0;
      tables.midway_byte3 = read24Snes(rom, p1 + 0x39) || 0;
      if (tables.midway_byte1 && tables.midway_byte2 && tables.midway_byte3) {
        tables.has_midway_hijack = true;
      }
    }
  }

  if (tables.has_midway_hijack) {
    let cand = 0;
    if (tables.midway_byte2 === tables.midway_byte1 + 0x200 &&
        tables.midway_byte3 === tables.midway_byte1 + 0x400) {
      cand = tables.midway_byte1 + 0x600;
    } else {
      cand = tables.midway_byte3 + 0x200;
    }
    if (cand && read8Snes(rom, cand) != null) {
      tables.midway_byte4 = cand;
      tables.has_midway_table4 = true;
    }
  }

  if (read8Snes(rom, snes(0x06, 0xfa00)) != null) tables.sec_byte6 = snes(0x06, 0xfa00);
  if (read8Snes(rom, snes(0x06, 0xfc00)) != null) tables.sec_byte7 = snes(0x06, 0xfc00);
  if (read8Snes(rom, snes(0x06, 0xfe00)) != null) tables.sec_byte8 = snes(0x06, 0xfe00);

  return { ok: true, tables };
}

module.exports = {
  lmResolveTables,
};
